src/driver-sinowealth.c

```diff
@@ -265,81 +458,144 @@ sinowealth_print_fw_version(struct ratbag_device *device) {
+		log_error(device->ratbag, "Couldn't read debounce time: %d\n", rc);
+		return -1;
+	}
+
+	/* Check if we got a response for correct command. */
+	if (buf[1] != SINOWEALTH_CMD_DEBOUNCE) {
+		log_error(device->ratbag, "Couldn't read debounce time, got result of command %#x instead\n", buf[1]);
+		return -1;
+	}
+
+	return buf[2] * 2;
+}
+
+/* Print angle snapping (Cal line) and lift-off distance (LOD) modes.
+ *
+ * This is only confirmed to work on G-Wolves Hati where the way with
+ * config report doesn't work. This does not work on Glorious Model O.
+ */
+static int
+sinowealth_print_long_lod_and_anglesnapping(struct ratbag_device *device)
+{
+	int rc = 0;
+
+	/* TODO: implement angle snapping and lift-off distance changing once we have an API for that.
+	 * To implement LOD changing here: set the third index to <whether you want LOD high or low> + 1.
+	 * To implement angle snapping toggling here: set the fourth index to 1 or 0 to enable or disable accordingly.
+	 */
+	uint8_t buf[SINOWEALTH_CMD_SIZE] = { SINOWEALTH_REPORT_ID_CMD, SINOWEALTH_CMD_LONG_ANGLESNAPPING_AND_LOD };
+	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CMD, buf, sizeof(buf));
+	if (rc != sizeof(buf)) {
+		log_error(device->ratbag, "Couldn't send LOD and angle snapping read command: %d\n", rc);
+		return -1;
+	}
+	rc = ratbag_hidraw_get_feature_report(device, SINOWEALTH_REPORT_ID_CMD, buf, sizeof(buf));
+	if (rc != sizeof(buf)) {
+		log_error(device->ratbag, "Couldn't read LOD and angle snapping: %d\n", rc);
+		return -1;
+	}
+
+	/* Check if we got a response for correct command. */
+	if (buf[1] != SINOWEALTH_CMD_LONG_ANGLESNAPPING_AND_LOD) {
+		log_error(device->ratbag, "Couldn't read LOD and angle snapping, got result of command %#x instead\n", buf[1]);
+		return -1;
+	}
+
+	log_info(device->ratbag, "LOD is high: %u\n", buf[2] - 1);
+	log_info(device->ratbag, "Angle snapping enabled: %u\n", buf[3]);
+
+	return 0;
+}
+
+/* Read configuration data from the mouse and save it in drv_data.
+ *
+ * @return 0 on success or an error code.
+ */
+static int
+sinowealth_read_raw_config(struct ratbag_device *device)
+{
 	int rc = 0;
 
 	struct sinowealth_data *drv_data = device->drv_data;
-	struct sinowealth_config_report *config = &drv_data->config;
+	struct sinowealth_config_report *config1 = &drv_data->configs[0];
 
-	uint8_t cmd[6] = {SINOWEALTH_REPORT_ID_CMD, SINOWEALTH_CMD_GET_CONFIG};
+	uint8_t cmd[SINOWEALTH_CMD_SIZE] = { SINOWEALTH_REPORT_ID_CMD, SINOWEALTH_CMD_GET_CONFIG };
 	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CMD, cmd, sizeof(cmd));
 	if (rc != sizeof(cmd)) {
 		log_error(device->ratbag, "Error while sending read config command: %d\n", rc);
 		return -1;
 	}
 
-	const char config_report_id = drv_data->is_long ? SINOWEALTH_REPORT_ID_CONFIG_LONG : SINOWEALTH_REPORT_ID_CONFIG;
+	const unsigned char config_report_id = drv_data->is_long ? SINOWEALTH_REPORT_ID_CONFIG_LONG : SINOWEALTH_REPORT_ID_CONFIG;
```

Comment:
do we get any benefit out of this being a char? Couldn't this just be a `uint8_t`? Or does the compiler complain here?

Arguably, we should change `ratbag_hidraw_get_feature_report()` to take the uint8_t instead of a char.

Reply:
I originally made this change because of the following clang-tidy warning about it's usage in `sinowealth_write_config`: `'signed char' to 'enum sinowealth_report_id' conversion; consider casting to 'unsigned char' first.`. 
Though you are right, I should've used `uint8_t` instead of `unsigned char`.
Fixed, thank you.